[RHEL8] Add support for multi NIC setup

enrico-usai opened the pull request:
### Description of changes
* Reverse logic in restart_network_service utility to have NetworkManager as default to support RedHat8
* Add kitchen tests for multi NIC setup
* Move metadata and network helpers to common cookbook
* Create resource for network_service
    * I defined two actions restart and reload, they do the same thing on alinux2, redhat8 and centos7 but on debian the reload action also applies the changes with netplan.

### Tests
* Tested on EC2 for Alinux2, RedHat, Ubuntu18, Centos7 and Ubuntu20 with p4d.24xlarge instance in a private subnet (see instructions in the new `kitchen.multinic.yaml` file.

Rhel8:
```
...
       Recipe: aws-parallelcluster-config::network_interfaces
         * log[macs: ["06:56:95:f1:1e:a1", "06:6b:75:40:89:ad", "06:c6:98:bb:c0:cb", "06:f5:cc:8f:85:df"]] action write
         * cookbook_file[configure_nw_interface.sh] action create (up to date)
         * log[device_number: 0, device_name: eth3, device_ip_address: 10.0.1.212] action write
         * execute[configure_nw_interface] action run
           - execute sh /tmp/configure_nw_interface.sh
         * log[device_number: 1, device_name: eth1, device_ip_address: 10.0.1.166] action write
         * execute[configure_nw_interface] action run
           - execute sh /tmp/configure_nw_interface.sh
         * log[device_number: 2, device_name: eth2, device_ip_address: 10.0.1.236] action write
         * execute[configure_nw_interface] action run
           - execute sh /tmp/configure_nw_interface.sh
         * log[device_number: 3, device_name: eth0, device_ip_address: 10.0.1.116] action write
         * execute[configure_nw_interface] action run
           - execute sh /tmp/configure_nw_interface.sh
         * network_service[Reload network configuration] action reload
           * service[NetworkManager] action restart
             - restart service service[NetworkManager]
...

  ✔  network_interfaces_configured: Check that network interfaces have been configured
     ✔  File /tmp/configure_nw_interface.sh is expected to exist
     ✔  File /tmp/configure_nw_interface.sh mode is expected to cmp == "0644"
     ✔  File /tmp/configure_nw_interface.sh owner is expected to eq "root"
     ✔  File /tmp/configure_nw_interface.sh group is expected to eq "root"
     ✔  File /tmp/configure_nw_interface.sh content is expected to match /^# Configure a specific Network Interface according to the OS/
     ✔  File /etc/sysconfig/network-scripts/ifcfg-eth1 is expected to exist
     ✔  File /etc/sysconfig/network-scripts/ifcfg-eth1 content is expected to match /^DEVICE=eth1/
     ✔  File /etc/sysconfig/network-scripts/route-eth1 is expected to exist
     ✔  File /etc/sysconfig/network-scripts/route-eth1 content is expected to match /^default via.*\seth1\stable\s100\d/
     ✔  File /etc/sysconfig/network-scripts/route-eth1 content is expected to match /^default via.*\seth1\smetric\s100\d/
     ✔  File /etc/sysconfig/network-scripts/route-eth1 content is expected to match /dev\seth1\stable\s100\d/
     ✔  File /etc/sysconfig/network-scripts/rule-eth1 is expected to exist
     ✔  File /etc/sysconfig/network-scripts/rule-eth1 content is expected to match /^from.*lookup\s100\d/
     ✔  Bash command ip a stdout is expected to match /inet.*eth1/
     ✔  File /etc/sysconfig/network-scripts/ifcfg-eth2 is expected to exist
     ✔  File /etc/sysconfig/network-scripts/ifcfg-eth2 content is expected to match /^DEVICE=eth2/
     ✔  File /etc/sysconfig/network-scripts/route-eth2 is expected to exist
     ✔  File /etc/sysconfig/network-scripts/route-eth2 content is expected to match /^default via.*\seth2\stable\s100\d/
     ✔  File /etc/sysconfig/network-scripts/route-eth2 content is expected to match /^default via.*\seth2\smetric\s100\d/
     ✔  File /etc/sysconfig/network-scripts/route-eth2 content is expected to match /dev\seth2\stable\s100\d/
     ✔  File /etc/sysconfig/network-scripts/rule-eth2 is expected to exist
     ✔  File /etc/sysconfig/network-scripts/rule-eth2 content is expected to match /^from.*lookup\s100\d/
     ✔  Bash command ip a stdout is expected to match /inet.*eth2/
     ✔  File /etc/sysconfig/network-scripts/ifcfg-eth3 is expected to exist
     ✔  File /etc/sysconfig/network-scripts/ifcfg-eth3 content is expected to match /^DEVICE=eth3/
     ✔  File /etc/sysconfig/network-scripts/route-eth3 is expected to exist
     ✔  File /etc/sysconfig/network-scripts/route-eth3 content is expected to match /^default via.*\seth3\stable\s100\d/
     ✔  File /etc/sysconfig/network-scripts/route-eth3 content is expected to match /^default via.*\seth3\smetric\s100\d/
     ✔  File /etc/sysconfig/network-scripts/route-eth3 content is expected to match /dev\seth3\stable\s100\d/
     ✔  File /etc/sysconfig/network-scripts/rule-eth3 is expected to exist
     ✔  File /etc/sysconfig/network-scripts/rule-eth3 content is expected to match /^from.*lookup\s100\d/
     ✔  Bash command ip a stdout is expected to match /inet.*eth3/
```

Ubuntu 20:
```
  ✔  network_interfaces_configured: Check that network interfaces have been configured
     ✔  File /tmp/configure_nw_interface.sh is expected to exist
     ✔  File /tmp/configure_nw_interface.sh mode is expected to cmp == "0644"
     ✔  File /tmp/configure_nw_interface.sh owner is expected to eq "root"
     ✔  File /tmp/configure_nw_interface.sh group is expected to eq "root"
     ✔  File /tmp/configure_nw_interface.sh content is expected to match /^# Configure a specific Network Interface according to the OS/
     ✔  File /etc/netplan/ens32.yaml is expected to exist
     ✔  File /etc/netplan/ens32.yaml content is expected to match /^    ens32:/
     ✔  File /etc/netplan/ens32.yaml content is expected to match /^\s+table:\s100\d/
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh is expected to exist
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh mode is expected to cmp == "0755"
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh owner is expected to eq "root"
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh group is expected to eq "root"
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh content is expected to match /^logger -t parallelcluster "Removing Automatic route/
     ✔  Bash command ip a stdout is expected to match /inet.*ens32/
     ✔  File /etc/netplan/ens33.yaml is expected to exist
     ✔  File /etc/netplan/ens33.yaml content is expected to match /^    ens33:/
     ✔  File /etc/netplan/ens33.yaml content is expected to match /^\s+table:\s100\d/
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh is expected to exist
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh mode is expected to cmp == "0755"
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh owner is expected to eq "root"
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh group is expected to eq "root"
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh content is expected to match /^logger -t parallelcluster "Removing Automatic route/
     ✔  Bash command ip a stdout is expected to match /inet.*ens33/
     ✔  File /etc/netplan/ens34.yaml is expected to exist
     ✔  File /etc/netplan/ens34.yaml content is expected to match /^    ens34:/
     ✔  File /etc/netplan/ens34.yaml content is expected to match /^\s+table:\s100\d/
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh is expected to exist
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh mode is expected to cmp == "0755"
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh owner is expected to eq "root"
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh group is expected to eq "root"
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh content is expected to match /^logger -t parallelcluster "Removing Automatic route/
     ✔  Bash command ip a stdout is expected to match /inet.*ens34/
     ✔  File /etc/netplan/ens35.yaml is expected to exist
     ✔  File /etc/netplan/ens35.yaml content is expected to match /^    ens35:/
     ✔  File /etc/netplan/ens35.yaml content is expected to match /^\s+table:\s100\d/
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh is expected to exist
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh mode is expected to cmp == "0755"
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh owner is expected to eq "root"
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh group is expected to eq "root"
     ✔  File /etc/networkd-dispatcher/routable.d/cleanup-routes.sh content is expected to match /^logger -t parallelcluster "Removing Automatic route/
     ✔  Bash command ip a stdout is expected to match /inet.*ens35/

Test Summary: 41 successful, 0 failures, 0 skipped
       Finished verifying <network-interfaces-ubuntu2004> (0m4.55s).
```

Ubuntu 18:
``` 
Test Summary: 41 successful, 0 failures, 0 skipped
       Finished verifying <network-interfaces-ubuntu1804> (0m3.54s).
```

alinux2
```
         * network_service[Reload network configuration] action reload
           * service[network] action restart
             - restart service service[network]
...
Test Summary: 28 successful, 0 failures, 0 skipped
       Finished verifying <network-interfaces-alinux2> (0m5.15s).
```

Centos7
```
Test Summary: 37 successful, 0 failures, 0 skipped
       Finished verifying <network-interfaces-centos7> (0m5.50s).
```

Commit:
[RHEL8] Add support for multi NIC setup

* Reverse logic in restart_network_service utility to have NetworkManager as default to support RedHat8
* Add kitchen tests for multi NIC setup
* Move metadata and network helpers to common cookbook

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Create resource for network_service

I defined two actions restart and reload, they do the same thing on alinux2, redhat8 and centos7
but on debian the reload action also applies the changes with netplan.

Signed-off-by: Enrico Usai <usai@amazon.com>

ermanno commented:
Is the reason we didn't test on Centos because the logic is basically the same as AL2?

enrico-usai commented:
> Is the reason we didn't test on Centos because the logic is basically the same as AL2?

Sorry I missed to update the description. I tested on Centos7 too.

ermanno commented on cookbooks/aws-parallelcluster-common/resources/network_service/partial/_restart_network_debian.rb:
```diff
@@ -0,0 +1,34 @@
+
+  service network_service_name.to_s do
+    action %i(restart)
+    ignore_failure true
```
Why is it safe to ignore failures here? I know that it's the same code that it was previously in the `libraries/helpers.rb`, but I was wondering what the rationale was.

enrico-usai replied:
I experienced that if you detach a networking interface and attach a new one you can have issues like:
``` 
Feb 27 11:54:53 ip-10-0-1-234.eu-west-1.compute.internal network[20916]: Bringing up interface eth3:  ERROR     : [/etc/sysconfig/network-scripts/ifup-eth] Device eth3 does not seem to be present, delaying initialization.
```
and in cases like this the restart of networking services exits with error code 1.

awslitvit commented on cookbooks/aws-parallelcluster-common/resources/network_service/network_service_ubuntu1804.rb:
```diff
@@ -0,0 +1,27 @@
+
+use 'partial/_restart_network_debian'
+
+action :restart do
+  action_restart_network
+end
+
+action :reload do
+  action_apply_network_changes
+end
```
Nitpick. Looks like common code which could go directly into `_restart_network_debian`.

enrico-usai replied:
you're right. I can move it.

enrico-usai replied:
Fixed.

awslitvit commented on cookbooks/aws-parallelcluster-common/resources/network_service/partial/_restart_network_alinux_centos.rb:
```diff
@@ -0,0 +1,24 @@
+  network_service_name = 'network'
+  Chef::Log.info("Restarting '#{network_service_name}' service, platform #{node['platform']} '#{node['platform_version']}'")
+
+  service network_service_name.to_s do
```
Nitpick: why `to_s`?

enrico-usai replied:
This was already there in the original code. I'll remove it.

enrico-usai replied:
Removed.

awslitvit commented on cookbooks/aws-parallelcluster-common/resources/network_service/partial/_restart_network_debian.rb:
```diff
@@ -0,0 +1,34 @@
+# See the License for the specific language governing permissions and limitations under the License.
+
+action :restart_network do
+  network_service_name = 'systemd-resolved'
```
Looks like this is the only difference between `alinux_centos` and `debian`. Could we define common code for `:restart_network` and define the service name for every OS separately?

enrico-usai replied:
Done

awslitvit commented on cookbooks/aws-parallelcluster-common/resources/network_service/partial/_restart_network_debian.rb:
```diff
@@ -0,0 +1,34 @@
+action :apply_network_changes do
+  action_restart_network
+
+  ruby_block "apply network configuration" do
```
Nitpick. Is there a reason not to use Chef [execute](https://docs.chef.io/resources/execute/) resource?

enrico-usai replied:
I just moved the already existing code. I can try to change it.

enrico-usai replied:
changed to use execute, tested on ubuntu20

awslitvit commented on cookbooks/aws-parallelcluster-common/resources/network_service/partial/_restart_network_debian.rb:
```diff
@@ -0,0 +1,34 @@
+
+action :restart_network do
+  network_service_name = 'systemd-resolved'
+  Chef::Log.info("Restarting '#{network_service_name}' service, platform #{node['platform']} '#{node['platform_version']}'")
```
Is there a reason not to use Chef [log](https://docs.chef.io/resources/log/) resource?

enrico-usai replied:
This was the original code, I just moved it. I can try to use the log resource.

enrico-usai replied:
Updated to use log resource and tested.

Commit:
Refactor network_service resource to avoid code duplication

restart action is in common for all OSes and the service name is defined with action_class
alinux2, centos7 and redhat8 have reload action in common,
ubuntu18 and ubuntu20 have both actions in common.

Signed-off-by: Enrico Usai <usai@amazon.com>

awslitvit commented on cookbooks/aws-parallelcluster-common/resources/network_service/network_service_ubuntu1804.rb:
```diff
@@ -0,0 +1,19 @@
+unified_mode true
+default_action :restart
+
+use 'partial/_restart_network_debian'
```
Could we change the name to `_network_service_debian`?

enrico-usai replied:
done

awslitvit commented on cookbooks/aws-parallelcluster-common/resources/network_service/partial/_restart_network_alinux_centos_redhat.rb:
```diff
@@ -0,0 +1,27 @@
+  log "Restarting '#{network_service_name}' service, platform #{node['platform']} '#{node['platform_version']}'"
+
+  service network_service_name do
+    action %i(restart)
```
Nitpick. Shouldn't `action :restart` be the same?

enrico-usai replied:
yes should be, I just moved previous code. I'll change it.

awslitvit commented on cookbooks/aws-parallelcluster-common/resources/network_service/partial/_restart_network_alinux_centos_redhat.rb:
```diff
@@ -0,0 +1,27 @@
+action :restart do
+  log "Restarting '#{network_service_name}' service, platform #{node['platform']} '#{node['platform_version']}'"
+
+  service network_service_name do
```
**Nitpick.** Looks like `:restart` action is common for all the OSes.
The only difference is that, in `:reload`, ubuntu OSes apply network configuration after restarting.
This is not immediately clear from the code.
I would suggest to have a common partial with `:restart` action (and maybe the default behavior for `:reload`) and a partial for `:reload` specific for ubuntu OSes.

enrico-usai replied:
Changed, good point, thanks!

awslitvit commented on cookbooks/aws-parallelcluster-common/resources/network_service/partial/_restart_network_debian.rb:
```diff
@@ -0,0 +1,38 @@
+end
+
+action :reload do
+  action_restart
```
Looks like restart wasn't performed for ubuntu OSes in the old code. Is it required?

enrico-usai replied:
This is not I mistakenly copied it on the debian branch too.

awslitvit commented on test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb:
```diff
@@ -0,0 +1,67 @@
+    if os_properties.debian_family?
+      describe file("/etc/netplan/#{device_name}.yaml") do
+        it { should exist }
+        its('content') { should match /^    #{device_name}:/ }
```
Nitpick. Is it critical to have exactly 4 whitespaces at the beginning of the line? Otherwise you could use `\s*` or `\s+`.

enrico-usai replied:
It's useful to check the spaces because there is a yaml and the device name is in a specific point of the tree, so the number of spaces matters. Anyway I can change it.

awslitvit commented on test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb:
```diff
@@ -0,0 +1,67 @@
+# This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+
+control 'network_interfaces_configured' do
```
Nitpick. I would split this control in:
- one that checks `*.sh` scripts have been written correctly
- the other checking actual network interfaces configuration

enrico-usai replied:
Done

awslitvit commented on test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb:
```diff
@@ -0,0 +1,67 @@
+        its('mode') { should cmp '0755' }
+        its('owner') { should eq 'root' }
+        its('group') { should eq 'root' }
+        its('content') { should match /^logger -t parallelcluster "Removing Automatic route/ }
```
Nitpick. I would check the file contains the line performing actual cleanup.

enrico-usai replied:
Done

Commit:
Split inspec tests into two controls

The first one verifies the script to manage the network interfaces is created correctly
the second one verifies it is executed correctly.

Signed-off-by: Enrico Usai <usai@amazon.com>

awslitvit commented on test/recipes/controls/aws_parallelcluster_config/network_interfaces_spec.rb:
```diff
@@ -0,0 +1,69 @@
+# This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+
+control 'network_interfaces_management_script_created' do
```
Nitpick. Shouldn't it be `network_interfaces_configuration_script_created`?

enrico-usai replied:
changed
